monadic-rx/shared/src/main/scala/mhtml/Optimizer.md

```diff
@@ -0,0 +1,227 @@
+
+* `dropIf`/`keepIf`
+
+    ```scala
+    val numbers: Rx[Int]
+    val odd: Rx[Int] = numbers.dropIf(_ % 2 == 0)(-1)
+    // numbers =>    0       0   3   4        5   6 ...
+    // odd1    =>       •-1      3            5     ...
+    // odd2    =>                       •-1   5     ...    
+    ```
+    
+We see that `odd1` emits -1, 3, 5 while `even2` emits -1, 5, so not everything can be 
+shared between the two streams. In other words, we cannot simply implement `odd2` by 
+"hooking" into `odd1` upon registration, since this would instead give the emission 
+sequence 3, 5. But we could do that after the 5 is emitted! 
+
+A similar argument can be used to show that merge is not always shareable but only 
+tail-shareable.
+    
+//TODO:    
+* `merge`
+    
+    ```scala
+    val r1: Rx[Int]
+    val r2: Rx[Int]
+    val merged: Rx[Int] = r1.merge(r2)
+    // r1      => 0    8     3 ...
+    // r2      => 1      4 3   ...
+    // merged1 => •0 1 8 4 3 3 ...
+    // merged2 =>   •1 8 4 3 3 ...
+    ```
+
+`foldp` is never-shareable, as can be seen from an extremely simple case:
+
+* `foldp`
+
+    ```scala
+    val numbers: Rx[Int]
+    val folded: Rx[Int] = numbers.foldp(0)(_ + _)
+    // numbers  => 1  2 1  1 3 ...
+    // folded1  => •1 3 4  5 8 ...
+    // folded2  =>      •1 2 5 ...
+    ```
+
+
+`flatMap` is dependently shareable; it "inherits" the shareability of whatever 
+is returned by its function. For instance, flatMapping on something that always 
+returns an Rx made of `map`, `zip`, or any other always-shareable `Rx` results in 
+something always shareable:
+
+* `flatMap` basic case (always shareable)
+
+    ```scala
+    val numbers: Rx[Int]
+    val evens: Rx[Int]
+    val odds: Rx[Int]
+    val flip: Rx[Int] = numbers.flatMap(x => if x % 2 == 0 evens else odds)
+    // numbers   => 1  2 1  4 3 ...
+    // evens     => 0  2 4  6 8 ...
+    // odds      => 1  3 5  7 9 ...
+    // flip1     => •1 2 5  6 9 ...
+    // flip2     =>      •5 6 9 ...
+    ```
+
+However, if we instead return something that is a `foldp`, we can no longer share:
+
+* `flatMap` never-shareable when returning a never-shareable `Rx`
+
+    ```scala
+    val numbers: Rx[Int]
+    def makeFolded: Rx[Int] = numbers.foldp(0)(_ + _)
+    val foldedFm: Rx[Int] = numbers.flatMap(x => makeFolded)
+    // numbers      => 1  2 1  1 3 ...
+    // foldedFm1    => •1 3 4  5 8 ...
+    // foldedFm2    =>      •1 2 5 ...  
+    ```
+
+
+* `flatMap` always-shareable when returning a previously defined `Rx`
+
+    ```scala
+    val numbers: Rx[Int]
+    val folded: Rx[Int] = numbers.foldp(0)(_ + _)
+    val foldedFm: Rx[Int] = numbers.flatMap(x => folded)
+    // numbers      => 1  2 1  1 3 ...
+    // folded       => •1 3 4  5 8 ...
+    // foldedFm1    => •1 3 4  5 8 ...
+    // foldedFm2    =>      •4 5 8 ...  
+    ```
+
+This suggests we should check to see if an `Rx` is defined (**how?**) as this 
```

Comment:
@OlivierBlanvillain Thanks for your previous comments, it is a big help. I'm curious now about how we can check this situation, where `folded` is already defined.

Also I'm still most immediately curious about how `merge` isn't always-shareable.

Reply:
Everything is RT, so having `folded` "already defined" or "defined on the fly" cannot make a different (by definition of RT)! I'm think if try it out in a test you will get the following outcome from this example:

```scala
val numbers: Rx[Int]
val folded: Rx[Int] = numbers.foldp(0)(_ + _)
val foldedFm: Rx[Int] = numbers.flatMap(x => folded)
// numbers      =>  1  2  1  1 3 ...
// folded       => •1  3  4  5 8 ...
// foldedFm1    => •1  3  4  5 8 ...
// foldedFm2    =>       •1  2 5 ...  
```

The problem about `merge` comes from the fact that is always uses the latest value from the left hand side Rx as it's first emitted value. For instance, in this example, merged2 is registered after r2 emitted 4 but will start with 8, the latest value from r1.

```scala
val r1: Rx[Int]
val r2: Rx[Int]
val merged: Rx[Int] = r1.merge(r2)
// r1      =>  0    8        3 ...
// r2      =>  1      4    3   ...
// merged1 =>    •0 8 4    3 3 ...
// merged2 =>           •8 3 3 ...
```

You can see that `8 3 3` is not a sub sequence of `0 8 4 3 3`, but `3 3` is, so merge is tail sharable.

Reply:
Ah right, of course - that sort of comes back to the whole point of all this.

For merge, this makes me realize/remember that an Rx must always have a value to emit (I think), so I need to fix some of the other examples a bit to reflect that.